crates/task-impls/src/transactions.rs

```diff
@@ -216,54 +216,61 @@ where
                 let txns = self.wait_for_transactions(parent_leaf).await?;
-                // TODO (Keyao) We encode the transactions and compute the VID disperse data twice,
-                // through `from_transactions` once, then `encode` and `disperse` again. This is
-                // because we need `disperse` for the `VidDisperseSend` event, which can't be
-                // retrieved by `from_transactions` directly.
-                // This duplication will be fixed when updating the data sent to the DA task, i.e.,
-                // in the `BlockReady` event.
-                // Relevant issue: https://github.com/EspressoSystems/HotShot/issues/2026.
-                let (payload, metadata) =
-                    <TYPES::BlockPayload as BlockPayload>::from_transactions(txns);
-                let encoded_txns = <TYPES::BlockPayload as BlockPayload>::encode(&payload);
-                // TODO <https://github.com/EspressoSystems/HotShot/issues/1686>
-                let srs = test_srs(NUM_STORAGE_NODES);
-                // TODO We are using constant numbers for now, but they will change as the quorum size
-                // changes.
-                // TODO <https://github.com/EspressoSystems/HotShot/issues/1693>
-                let vid = VidScheme::new(NUM_CHUNKS, NUM_STORAGE_NODES, &srs).unwrap();
-                let vid_disperse = vid.disperse(encoded_txns).unwrap();
-
-                // TODO never clone a block
-                // https://github.com/EspressoSystems/HotShot/issues/1858
-                self.event_stream
-                    .publish(HotShotEvent::BlockReady(
-                        payload.clone(),
-                        metadata,
-                        view + 1,
-                    ))
-                    .await;
-
-                // TODO (Keyao) Determine and update where to publish VidDisperseSend.
-                // <https://github.com/EspressoSystems/HotShot/issues/1817>
-                debug!("publishing VID disperse for view {}", *view + 1);
-                self.event_stream
-                    .publish(HotShotEvent::VidDisperseSend(
-                        Proposal {
-                            data: VidDisperse {
-                                view_number: view + 1,
-                                payload_commitment: payload.commit(),
-                                shares: vid_disperse.shares,
-                                common: vid_disperse.common,
-                            },
-                            // TODO (Keyao) This is also signed in DA task.
-                            signature: self
-                                .quorum_exchange
-                                .sign_payload_commitment(payload.commit()),
-                        },
-                        self.quorum_exchange.public_key().clone(),
-                    ))
-                    .await;
+                match <TYPES::BlockPayload as BlockPayload>::from_transactions(txns) {
```

Comment:
Nit: I feel like we can clean up this nesting slightly by returning early on errors. Like
```rust
let x = match fallible_function() {
  Ok(x) => x,
  Err(e) => {
    error!("something");
    return None;
  }
};
let y = match another_fallible_function(x) {
  ...
};
```